allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py

```diff
@@ -0,0 +1,170 @@
+        corresponding embeddings will be removed from the return values.
+
+        Warning: This only removes a single start and single end token!
+    requires_grad : ``bool``, optional (default: False)
+        If True, compute gradient of bidirectional language model parameters for fine tuning.
+    """
+    def __init__(self,
+                 archive_file: str,
+                 dropout: float = None,
+                 bos_eos_tokens: Tuple[str, str] = ("<S>", "</S>"),
+                 remove_bos_eos: bool = True,
+                 requires_grad: bool = False) -> None:
+        super().__init__()
+
+        overrides = {
+                "model": {
+                        "contextualizer": {
+                                "return_all_layers": True
+                        }
+                }
+        }
+
+        # Load LM and the associated config.
+        archive = load_archive(archive_file, overrides=json.dumps(overrides))
+        self._lm: BidirectionalLanguageModel = archive.model
+        # pylint: disable=protected-access
+        self._lm._softmax_loss = None
+        config = archive.config
+        dict_config = config.as_dict(quiet=True)
+
+        # Extract the name of the tokens that the LM was trained on.
+        token_names = list(dict_config["model"]["text_field_embedder"]["token_embedders"].keys())
+        if len(token_names) != 1:
+            # A TokenEmbedder embeds a single set of indices. If the LM was trained with multiple embedders, and
+            # thus indices, we won't be able to embed a single set of indices here.
+            #
+            # Note: We only care about embedded indices. This does not include "tokens" which is just used to
+            # compute the loss in BidirectionalLanguageModel.
+            raise RuntimeError(f"LM from {archive_file} trained with multiple embedders!")
+        self._token_name = token_names[0]
+
+        # TODO(brendanr): Remove this hack once we have proven we can train without using the
+        # ELMoTokenCharactersIndexer. For the equivalent hack in the ELMo embedder see
+        # https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L590.
+        if bos_eos_tokens:
+            dataset_reader_config = config.get("dataset_reader")
+            if dataset_reader_config.get("type") == "multiprocess":
+                dataset_reader_config = dataset_reader_config.get("base_reader")
+            token_indexer_config = dataset_reader_config.get("token_indexers").get(self._token_name)
+            token_indexer: TokenIndexer = TokenIndexer.from_params(token_indexer_config)
+            token_list = [Token(token) for token in bos_eos_tokens]
+            bos_eos_indices = token_indexer.tokens_to_indices(token_list, self._lm.vocab, "key")["key"]
+            self._bos_indices = torch.Tensor(bos_eos_indices[0])
+            self._eos_indices = torch.Tensor(bos_eos_indices[1])
```

Comment:
wouldn't this whole section be simpler if you just called `Vocabulary.get_token_index`  for the bos and eos tokens? then you wouldn't even need the token indexer?

Reply:
+1

Reply:
I see, the wordpiece and openai indexers also have their own id mappings, but they add them to the vocabulary, so you can do things like this. whereas it looks like the ELMo indexer doesn't. probably it should, but definitely not as part of this PR. can you just add a TODO for that?

Reply:
Good call, thanks for the pointer! TODO added.